Wrapper function for functions with similar functionality

5saviahv opened the pull request:
Wrapper function "_matcher" for functions with similar functionality.

* It changes .parents function sorting (test shows strange sorting order)
* it changes .siblings function matched elements (it matches same elements what jQuery [first example on their website] )  

other functions work like before, simply repeating parts of the functions are kept in "wrapper" function

fb55 commented:
Hi @5saviahv, this is an awesome idea, thanks for opening this PR! I just started a new job, so might need some time to review this properly – sorry about that. Also no need to rebase the PR (that is great, but GitHub can deal with it).

5saviahv commented:
Thanks for telling, take your time - in jQuery there are many such functions, it reduces repeating same constructs over and over.

5saviahv commented:
- added fixture `eleven` so order of elements will be visible also it is separated by _LF_ so tests have to deal with text nodes as well.
- added some comments to function

fb55 commented on lib/api/traversing.js:
```diff
@@ -12,6 +12,46 @@ var uniqueSort = require('htmlparser2').DomUtils.uniqueSort;
+      if (typeof selector === 'string') {
+        matched = select.filter(selector, matched, this.options);
+      } else {
+        matched = uniqueSort(
```
None of the previous methods seem to be using `uniqueSort`, which will lead to a slowdown here. Maybe uniqueness could be an option to this function, and done using a linear search as before?

5saviahv replied:
Yes, `prev*` and `next*` functions wont need sorting, but since `select.filter` uses same sorting internally it messes up order, if filter is used.

5saviahv replied:
also function `children` is unique but for all other functions uniqueSort should only used if original cheerio object length would be more than 1.

Since if you take `parents` from one tag each element would be unique and reverse sorted, but from two and up elements start to differ and one point start to be duplicate, also order may be harder to control. Same with most other functions.

fb55 reviewed (commented):
This is a great idea! Right now, it will lead to a slowdown due to `uniqueSort` being used quite a bit, when some of the results were taken without a change before. I've left some comments, let me know if they are useful.

5saviahv commented on lib/api/traversing.js:
```diff
@@ -12,6 +12,46 @@ var uniqueSort = require('htmlparser2').DomUtils.uniqueSort;
 
+function _matcher(fn, reverse) {
+  // customized Map, discards null elements
+  function matchMap(elems, cb, arg) {
```
yeah I had idea using it with Until functions but I turned down that idea.

5saviahv replied:
Done

5saviahv commented:
Sry, `while loop` in `siblings` function kind of broke thing (first element was discarded) and I wasnt able revert it so I overwrite it.

5saviahv commented on lib/api/traversing.js:
```diff
@@ -12,6 +12,46 @@ var uniqueSort = require('htmlparser2').DomUtils.uniqueSort;
+      }
+
+      // Reverse order for parents* and prev-derivatives
+      if (reverse) {
```
It would be **idea**, but sorting in select.filter would make it meaningless.

5saviahv commented on lib/api/traversing.js:
```diff
@@ -12,6 +12,46 @@ var uniqueSort = require('htmlparser2').DomUtils.uniqueSort;
+    var i = 0;
+    var ret = [];
+    for (; i < len; i++) {
+      value = cb(elems[i], i, arg);
```
OK, Done

5saviahv commented:
I added [Cheerio-Select #7](https://github.com/cheeriojs/cheerio-select/issues/7) and it is related to this. We can later come back and change how `uniqueSort` works.

5saviahv commented on lib/api/traversing.js:
```diff
@@ -471,25 +438,22 @@ exports.prevUntil = function (selector, filterSelector) {
+exports.siblings = _matcher(function (elem) {
+  var node = elem.parent && elem.parent.firstChild;
+  var matched = [];
+  for (; node; node = node.next) {
```
first element would be lost

5saviahv commented on lib/api/traversing.js:
```diff
@@ -233,30 +246,17 @@ exports.closest = function (selector) {
-};
+exports.next = _matcher(function (elem) {
+  // eslint-disable-next-line no-empty
+  while ((elem = elem.next) && !isTag(elem)) {}
```
I didnt know while can use this way.

fb55 reviewed (commented):
> I added https://github.com/cheeriojs/cheerio-select/issues/7 and it is related to this. We can later come back and change how uniqueSort works.

I have updated cheerio-select's `filter` method to no longer change the order of the input. That hopefully (?) removes the need for sorting in here entirely.

fb55 commented on lib/api/traversing.js:
```diff
@@ -12,6 +12,63 @@ var uniqueSort = require('htmlparser2').DomUtils.uniqueSort;
+        ret.push(value);
+      }
+    }
+    return Array.prototype.concat.apply([], ret);
```
Or is this flattening the array?
